### PR TITLE
[DR-1680] Update webapp template calls

### PR DESCRIPTION
**Pull request by cristiangiagante**
## Background
  
Update webapi template calls to use new methods to get template's body
  
## Changes done
  
Add new methods on template api service

Call to template body service methods on template controller

Add tests to edit and create templates
  
## Pending to be done
  
~Tests~

~Error handling~

## Note

We will do the error handling when EditTemplateBody fails later.

### CODE REVIEW

File: src/wwwroot/services/templates.js [hunk 1]
```diff
@@ -69,14 +83,34 @@
...
+          return result.createdResourceId;
+        });
       } else {
+        var body = request.data.body;
+        // Prevent to update body using deprecated property
+        request.data.body = "";
         request.actionDescription = '"Editing a template"',
         request.method = 'PUT',
         request.url = RELAY_CONFIG.baseUrl + '/accounts/' + accountId + '/templates/' + model.id;
+        return editTemplateBody(accountId, model.id, request.data.name, body).then(function (r) {
```
Comment (andresmoschini) [inline src/wwwroot/services/templates.js]: What is _r_?

File: src/wwwroot/services/templates.js [hunk 2]
```diff
@@ -69,14 +83,34 @@
...
+          return $http(request).then(function (result) {
```
Comment (andresmoschini) [inline src/wwwroot/services/templates.js]: Since you are refactoring this method, maybe it is a good idea refactoring it totally in place of keeping those weird things to reuse some code.

File: src/wwwroot/services/templates.js (hunk 1 above)
Comment (andresmoschini) [inline src/wwwroot/services/templates.js]: Did you think how to deal with errors? For example, what happen if one request works and the another fails?

**Comment by andresmoschini** [comment]
> What do you think about create test to template controller? I guess it can be usefull with the new template editor

I think that we should add those tests.

File: src/wwwroot/controllers/TemplateCtrl.js [hunk 3]
```diff
@@ -39,19 +39,25 @@
...
-          $scope.template.subject = result.subject;
-          $scope.template.content = result.body;
-          $scope.template.name = result.name;
-          $scope.template.id = result.id;
-        }).catch(function (error) {
-          // TODO: do something with the error
-        }).finally(function () {
-          $scope.loadInProgress = false;
-        });
+      templates.getTemplateBody(templateId)
```
[the remark this answers is not in the document]
Comment (cristiangiagante) [reply src/wwwroot/controllers/TemplateCtrl.js]: > Why did you get template body first and then template? I understand that it works, but it is logically weird.

I do that to ensure the body response before the template request is finished to show all at the same time when the load spinner but yes, It's weird. I will refact it.

> Also, I remember that we had a kind of cache, is that right? or it was related to another thing?

I don't found anything related to a cache in the template ctrl or template service.

> I think that you are not dealing with finally in the right place.

I will check it.

> Another thing, with this approach, we cannot take advantage of API links to get the body, I am not really sure, but we could move that responsability to service.

Could you please explain me in more details? templates.getTemplateBody lives in template service.

File: src/wwwroot/services/templates.js [hunk 4]
```diff
@@ -53,11 +54,24 @@
...
+    function getTemplateBody(id) {
+      var accountId = auth.getAccountId();
+
+      return $http({
+        actionDescription: 'action_templates_getting_body',
+        method: 'GET',
+        url: RELAY_CONFIG.baseUrl + '/accounts/' + accountId + '/templates/' + id + '/body'
+      })
+        .then(function (result) {
+          return result.data;
```
[the remark this answers is not in the document]
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: Ok

File: src/wwwroot/services/templates.js [hunk 5]
```diff
@@ -53,11 +54,24 @@
...
 
+    function getTemplateBody(id) {
+      var accountId = auth.getAccountId();
+
+      return $http({
+        actionDescription: 'action_templates_getting_body',
+        method: 'GET',
+        url: RELAY_CONFIG.baseUrl + '/accounts/' + accountId + '/templates/' + id + '/body'
+      })
+        .then(function (result) {
```
[the remark this answers is not in the document]
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: I will fix it.

File: src/wwwroot/services/templates.js [hunk 6]
```diff
@@ -69,14 +83,34 @@
...
         request.method = 'POST',
         request.url = RELAY_CONFIG.baseUrl + '/accounts/' + accountId + '/templates';
+
+        return $http(request).then(function (result) {
+          return result.createdResourceId;
+        });
       } else {
+        var body = request.data.body;
+        // Prevent to update body using deprecated property
+        request.data.body = "";
```
[the remark this answers is not in the document]
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: It is, to not override values with the old api calls. I will add PATCH in the comment line.

[the remark this answers is not in the document]
Comment (andresmoschini) [reply src/wwwroot/services/templates.js]: I understand, but there should be better alternatives. What do you think about an explicit mapping like in getTemplate?

File: src/wwwroot/controllers/TemplateCtrl.js (hunk 3 above)
Comment (andresmoschini) [inline src/wwwroot/controllers/TemplateCtrl.js]: > I don't found anything related to a cache in the template ctrl or template service.

👍

Comment (andresmoschini) [inline src/wwwroot/controllers/TemplateCtrl.js]: >  > Another thing, with this approach, we cannot take advantage of API links to get the body, I am not really sure, but we could move that responsability to service.

> Could you please explain me in more details? templates.getTemplateBody lives in template service.

I am talking about moving orchestration responsibility to the service. For example, adding a new method like _getTemplateWithBody_.

File: src/wwwroot/services/templates.js [hunk 7]
```diff
@@ -69,14 +83,34 @@
         request.actionDescription = 'action_templates_creating',
         request.method = 'POST',
         request.url = RELAY_CONFIG.baseUrl + '/accounts/' + accountId + '/templates';
+
+        return $http(request).then(function (result) {
+          return result.createdResourceId;
+        });
```
[the remark this answers is not in the document]
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: Oh sorry I used the deprecated property body in template creation I need to update the template body with the new api call.

File: src/wwwroot/services/templates.js (hunk 2 above)
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: 👍

File: src/wwwroot/services/templates.js (hunk 1 above)
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: I will investigate it, thanks

File: src/wwwroot/services/templates.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: 👍

File: src/wwwroot/services/templates.js (hunk 1 above)
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: the result of the promise

Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: What do you think about to create a template model backup to rollback if the second request fails?

File: src/wwwroot/controllers/TemplateCtrl.js [hunk 8]
```diff
@@ -39,19 +39,16 @@
...
-      templates.getTemplate(templateId)
-        .then(function (result) {
-          $scope.template.fromName = result.from_name;
-          $scope.template.fromEmail = result.from_email;
-          $scope.template.subject = result.subject;
-          $scope.template.content = result.body;
-          $scope.template.name = result.name;
-          $scope.template.id = result.id;
-        }).catch(function (error) {
-          // TODO: do something with the error
```
Comment (andresmoschini) [inline src/wwwroot/controllers/TemplateCtrl.js]: Did you test what happens when you remove this catch?

File: src/wwwroot/services/templates.js [hunk 9]
```diff
@@ -19,7 +19,8 @@
       getAllData: getAllData,
       getTemplate: getTemplate,
```
Comment (andresmoschini) [inline src/wwwroot/services/templates.js]: Maybe this method could be removed since it should not be being used anymore.

File: src/wwwroot/services/templates.js [hunk 10]
```diff
@@ -52,33 +53,112 @@
         actionDescription: 'action_templates_getting',
         method: 'GET',
         url: RELAY_CONFIG.baseUrl + '/accounts/' + accountId + '/templates/' + id
-      })
-          .then(function (result) {
-            return result.data;
-          });
+      }).then(function (result) {
+        return extractData(result);
+      });
```
Comment (andresmoschini) [inline src/wwwroot/services/templates.js]: it could be rewritten as:

```
}).then(extractData);
```

Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: 👍

File: src/wwwroot/services/templates.js (hunk 9 above)
Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: Yes, I forget it. thx

File: src/wwwroot/controllers/TemplateCtrl.js (hunk 8 above)
Comment (cristiangiagante) [reply src/wwwroot/controllers/TemplateCtrl.js]: Yes, the modal window error is shown

Comment (andresmoschini) [reply src/wwwroot/controllers/TemplateCtrl.js]: ❤️

**Comment by cristiangiagante** [comment]
@mvillaseco @andresmoschini I add some tests to template page, could you review please? thx

File: src/spec/page-objects/template-page.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
+class TemplatePage {
+  constructor() {
+    this._templateNameLabel = $("form .templates--information .information-fields label[for='uName'");
+    this._templateFromEmailLabel = $("form .templates--information .information-fields label[for='uFromEmail'");
+    this._templateFromNameLabel = $("form .templates--information .information-fields label[for='uFromName'");
+    this._templateSubjectLabel = $("form .templates--information .information-fields label[for='uSubject'");
```
Comment (andresmoschini) [inline src/spec/page-objects/template-page.js]: Is it not so much specific? for example, if _templates--information_ is the form's class,  it should be not necessary using _information-fields_, right?

File: src/spec/template_spec.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    setupTemplateResponse();
+
+    var templatePage = new TemplatePage();
+    var expectedTitleText = "Template Info";
+    var expectedTemplateNameText = "Template name";
+    var expectedFromEmailText = "From Email";
+    var expectedFromNameText = "From Name";
+    var expectedSubjectText = "Subject";
+    var expectedTemplateEditorTitle = "Template Editor";
+    var expectedTemplateEditorDescription = "Here's where you can create and edit your custom template.";
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: I think that testing literals is not a good idea.

File: src/spec/template_spec.js [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+    var body = "Test content body";
+
+    var templatePage = new TemplatePage();
+
+    // Act
+    templatePage.FillTemplateBodySection(body);
+    templatePage.clickSaveTemplateButton();
+    
+    // Assert
+    expect(browser.getCurrentUrl()).not.toContain('templates/');
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: > _should save template in edit mode and redirect to templates list_
> _expect(browser.getCurrentUrl()).not.toContain('templates/');_

Does not _templates list_ contains _templates/_ in the URL? what is the final URL? why are you not testing for it?

File: src/spec/template_spec.js [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: _should save template in edit mode and redirect to templates list_

I do not see where are you testing if it saves the template.

File: src/wwwroot/services/templates.js [hunk 15]
```diff
@@ -17,9 +17,9 @@
 
     var templatesService = {
       getAllData: getAllData,
-      getTemplate: getTemplate,
+      getTemplateWithBody: getTemplateWithBody,
       save: save,
-      deleteTemplate: deleteTemplate
+      deleteTemplate: deleteTemplate,
```
Comment (mvillaseco) [inline src/wwwroot/services/templates.js]: extra comma?

File: src/spec/template_spec.js [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+    templatePage.fillTemplateNameInput(templateName);
+    templatePage.fillTemplateFromEmailInput(templateFromEmail);
+    templatePage.fillTemplateFromNameinput(templateFromName);
+    templatePage.fillTemplateSubjectInput(templateSubject);
+    templatePage.FillTemplateBodySection(templateBody);
+    templatePage.clickSaveTemplateButton();
+    
+    // Assert
+    expect(browser.getCurrentUrl()).not.toContain('templates/new');
+    expect(templatePage.getLastestSquareGuidCreatedText()).toBe(expectedGuid);
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: > _should create a new template_

What does it mean? I did not see that a new template is created in these assertions.

File: src/wwwroot/services/templates.js (hunk 15 above)
Comment (andresmoschini) [inline src/wwwroot/services/templates.js]: Could you remove the trailing comma?

Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: 👍

Comment (cristiangiagante) [reply src/wwwroot/services/templates.js]: 👍

File: src/spec/page-objects/template-page.js (hunk 11 above)
Comment (cristiangiagante) [reply src/spec/page-objects/template-page.js]: Yes it is not necessary, I will remove it to be more flexible.

File: src/spec/template_spec.js (hunk 12 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: @mvillaseco do you know if it's possible to include en-translation.js / es-translation in tests?

Comment (cristiangiagante) [reply src/spec/template_spec.js]: @andresmoschini do you think this test is necessary? I don't found a better approach to test it.

Comment (cristiangiagante) [reply src/spec/template_spec.js]: I will remove this test

File: src/spec/page-objects/template-page.js [hunk 17]
```diff
@@ -0,0 +1,103 @@
+class TemplatePage {
+  constructor() {
+    this._templateNameLabel = $("form .templates--information label[for='uName'");
+    this._templateFromEmailLabel = $("form .templates--information label[for='uFromEmail'");
+    this._templateFromNameLabel = $("form .templates--information label[for='uFromName'");
+    this._templateSubjectLabel = $("form .templates--information label[for='uSubject'");
+    this._templateEditorDescription = $("form .templates--edition label[for='uContent']");
+    this._templatePageTitle = $('form .templates--information h2');
```
Comment (andresmoschini) [inline src/spec/page-objects/template-page.js]: I would remove this kind of elements, we do not need to read a title, right?

File: src/spec/template_spec.js [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+
+      $httpBackend.whenPOST(/\/accounts\/[\d|-]*\/templates/).respond(500, {
+        "createdResourceId": "1169ab26-78dc-4474-a8c2-e1f02f6c05ed",
+        "message": "Template creation fail",
+        "_links": []
+      });
+    }));
+  }
+
+  function setupTemplateMenuSuccessResponse() {
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: Menu?

File: src/spec/template_spec.js [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+
+    // Act
+    templatePage.fillTemplateBodySection(body);
+    templatePage.clickSaveTemplateButton();
+    
+    // Assert
+    expect(browser.getCurrentUrl()).toEqual(browser.baseUrl + '/#/templates');
+  });
+
+  it('should save template in edit mode and fail', () => {
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: What do you think about: _It should show error and do not redirect when save fails_ ?

File: src/spec/template_spec.js [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+    setupTemplateMenuSuccessResponse();
+    var body = "Test content body";
+    var templatePage = new TemplatePage();
+
+    // Act
+    templatePage.fillTemplateBodySection(body);
+    templatePage.clickSaveTemplateButton();
+    
+    // Assert
+    expect(templatePage.isErrorModalDisplayed()).toBeTruthy();
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: Could you verify that the URL did not change?

File: src/spec/template_spec.js [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+  });
+
+  it('should create a new template and fail', () => {
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: Could you rephrase the based on my previous comment?

File: src/spec/template_spec.js [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+    templatePage.fillTemplateFromNameinput(templateFromName);
+    templatePage.fillTemplateSubjectInput(templateSubject);
+    templatePage.fillTemplateBodySection(templateBody);
+    templatePage.clickSaveTemplateButton();
+    
+    // Assert
+    expect(templatePage.isErrorModalDisplayed()).toBeTruthy();
+  });
+
+  it('should create a new template and redirect to template list', () => {
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: What do you think about: _It should redirect to template list after successfully save_ ?

File: src/spec/template_spec.js (hunk 18 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: The list of templates, yeah is really ugly name, I will change it 👍

File: src/spec/template_spec.js [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+    templatePage.fillTemplateFromNameinput(templateFromName);
+    templatePage.fillTemplateSubjectInput(templateSubject);
+    templatePage.fillTemplateBodySection(templateBody);
+    templatePage.clickSaveTemplateButton();
+    
+    // Assert
+    expect(browser.getCurrentUrl()).toEqual(browser.baseUrl + '/#/templates');
+  });
+
+  it('should not allow to create a new template if the form is not completed', () => {
```
Comment (andresmoschini) [inline src/spec/template_spec.js]: It is more common the usage of _when_ in place of _if_

File: src/wwwroot/controllers/TemplateCtrl.js [hunk 24]
```diff
@@ -39,19 +39,16 @@
...
-        }).finally(function () {
-          $scope.loadInProgress = false;
-        });
+      templates.getTemplateWithBody(templateId).then(function (result) {
```
Comment (andresmoschini) [inline src/wwwroot/controllers/TemplateCtrl.js]: What do you think about return the promise?

```
return templates.getTemplateWithBody(templateId).then(function (result) {
```

File: src/spec/template_spec.js (hunk 19 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: Thanks for the help I take it is more detailed.

**Comment by andresmoschini** [comment]
Really nice work! Maybe you could get some ideas about test naming from https://github.com/aspnet/Home/wiki/Engineering-guidelines or maybe Jasmine or other frameworks...

File: src/spec/page-objects/template-page.js (hunk 17 above)
Comment (cristiangiagante) [reply src/spec/page-objects/template-page.js]: 👍

File: src/spec/template_spec.js (hunk 21 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: 👍

File: src/spec/template_spec.js (hunk 22 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: 👍

File: src/spec/template_spec.js (hunk 23 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: 👍  thx

File: src/wwwroot/controllers/TemplateCtrl.js (hunk 24 above)
Comment (cristiangiagante) [reply src/wwwroot/controllers/TemplateCtrl.js]: I like to be consistent with other methods like save(), they don't return promises, and the parent method is called only in a private scope at the controller.

File: src/spec/template_spec.js (hunk 20 above)
Comment (cristiangiagante) [reply src/spec/template_spec.js]: 👍

File: src/wwwroot/controllers/TemplateCtrl.js (hunk 24 above)
Comment (andresmoschini) [reply src/wwwroot/controllers/TemplateCtrl.js]: It is not important at this moment, I also like consistency, but I prefer moving the consistency to promises approach.

It is powerful and cheap, I mean, if you return the promise, you could take advantage of it, or simply ignore it. If you do not return the promise you _have to_ ignore it.

File: src/wwwroot/controllers/TemplateCtrl.js [hunk 25]
```diff
@@ -57,7 +57,7 @@
       }
 
       $scope.saveInProgress = true;
-      templates.save({
+      return templates.save({
```
Comment (andresmoschini) [inline src/wwwroot/controllers/TemplateCtrl.js]: ❤️ 
Although it was not necessary I was expressing my opinion, not asking you to apply a change right now. but thanks.

Comment (cristiangiagante) [reply src/wwwroot/controllers/TemplateCtrl.js]: I like your opinion about why is good idea implement it 👍
